source/link/linker.cpp

```diff
@@ -0,0 +1,755 @@
+        return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                          SPV_ERROR_INVALID_BINARY)
+               << "Decorations mismatch between imported variable/function %" << i.first.id
+               << " and exported variable/function %" << i.second.id << ".";
+    if (!haveIdsSimilarDecorations(i.first.typeId, modules[i.first.moduleIndex],
+                                   i.second.typeId, modules[i.second.moduleIndex]))
+        return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                          SPV_ERROR_INVALID_BINARY)
+               << "Decorations mismatch between the type of imported variable/function %" << i.first.id
+               << " and the type of exported variable/function %" << i.second.id << ".";
+    for (uint32_t j = 0u; j < i.first.parametersIds.size(); ++j)
+      if (!haveIdsSimilarDecorations(i.first.parametersIds[j], modules[i.first.moduleIndex],
+                                     i.second.parametersIds[j], modules[i.second.moduleIndex]))
+          return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                            SPV_ERROR_INVALID_BINARY)
+                 << "Decorations mismatch between imported function %" << i.first.id << "'s"
+                 << " and exported function %" << i.second.id << "'s " << j << "th parameter.";
+  }
+
+  // TODO(pierremoreau): Remove import types
+  // Similar to the testing: traverse bottom-top
+  // But make sure to store the ID of removed types, to do the replacement
+  // later.
+
+  // Remove import decorations
+  for (const auto& i : linkingsToDo) {
+    removeDecorationsFor(i.first.id, modules[i.first.moduleIndex]);
+    removeDecorationsFor(i.first.typeId, modules[i.first.moduleIndex]);
+    for (const auto j : i.first.parametersIds)
+      removeDecorationsFor(j, modules[i.first.moduleIndex]);
+  }
+
+  // Remove prototypes of imported functions
+  for (const auto& i : linkingsToDo) {
+    for (auto j = modules[i.first.moduleIndex]->begin();
+        j != modules[i.first.moduleIndex]->end();)
+      j = (j->result_id() == i.first.id) ? j.Erase() : ++j;
+  }
+
+  // Remove declarations of imported variables
+  for (const auto& i : linkingsToDo) {
+    for (auto j = modules[i.first.moduleIndex]->types_values_begin();
+        j != modules[i.first.moduleIndex]->types_values_end();)
+      j = (j->result_id() == i.first.id) ? j.Erase() : ++j;
+  }
+
+  // Phase 2: Merge all the binaries into a single one.
+  auto linkedModule = MakeUnique<Module>();
+  libspirv::AssemblyGrammar grammar(impl_->context);
+  spv_result_t res =
+      MergeModules(modules, grammar, impl_->context->consumer, linkedModule);
+  if (res != SPV_SUCCESS) return res;
+
+  // Rematch import variables/functions to export variables/functions
+  linkedModule->ForEachInst([&linkingsToDo](Instruction* insn) {
```

Comment:
Suggest you use ReplaceAllUsesWith  ?  But that assumes you have a DefUseManager.  You could create one right now.